### PR TITLE
filterdb: Implement PurgeFilters.

### PR DESCRIPTION
**Pull request by roeierez**
This PR adds the ability to purge the filters from the db. Neutrino has the `PersistToDisk` query option which is useful for background sync in mobile applications. These apps can set a background job to fetch the filters periodically, preparing the data for the app when it starts by the user, which achieves a fast sync time and immediate readiness for payment.
This change adds a way to purge these filters and enables the app to implement its own eviction strategy preventing the db from growing boundlessly.

### CODE REVIEW

**Comment by Crypt-iQ** [comment]
I like this change, but do you notice a change in the size of the db when running this? `bbolt` doesn't shrink when deleting things (keeps the freed pages on disk still). May be useful if/when neutrino migrates to an abstracted k/v store that supports things besides `bbolt`.

**Comment by roeierez** [comment]
The bolt db file size indeed doesn't change but with this change the app layer at least can make sure it's eviction strategy prevents the db from growing too much.
As far as I know the only way to shrink the db is to run compact and this is also an option (that should be done with care) for applications that already have a big db file.

**Comment by Crypt-iQ** [comment]
> The bolt db file size indeed doesn't change but with this change the app layer at least can make sure it's eviction strategy prevents the db from growing too much.
> As far as I know the only way to shrink the db is to run compact and this is also an option (that should be done with care) for applications that already have a big db file.

Ah I see, I'll take a look at the code :)

File: filterdb/db.go [hunk 1]
```diff
@@ -51,6 +51,9 @@ type FilterDatabase interface {
 	// target block hash cannot be found, then ErrFilterNotFound is to be
 	// returned.
 	FetchFilter(*chainhash.Hash, FilterType) (*gcs.Filter, error)
+
+	// PurgeFilters purge all filters with a given type from persistent storage
+	PurgeFilters(FilterType) error
```
Comment (halseth) [inline filterdb/db.go]: dead code?

Comment (roeierez) [reply filterdb/db.go]: Not used within neutrino but intended to be used by apps that use neutrino as a chain service.
For example, we use a background job that run neutrino  to persist the filters and this gives us a way to purge the filters when not needed.

File: filterdb/db.go [hunk 2]
```diff
@@ -51,6 +51,9 @@ type FilterDatabase interface {
 	// target block hash cannot be found, then ErrFilterNotFound is to be
 	// returned.
 	FetchFilter(*chainhash.Hash, FilterType) (*gcs.Filter, error)
+
+	// PurgeFilters purge all filters with a given type from persistent storage
```
Comment (halseth) [inline filterdb/db.go]: nit: missing period

File: filterdb/db.go [hunk 3]
```diff
@@ -107,6 +110,28 @@ func New(db walletdb.DB, params chaincfg.Params) (*FilterStore, error) {
 	}, nil
 }
 
+// PurgeFilters purge all filters with a given type from persistent storage
+// NOTE: This method is a part of the FilterDatabase interface.
```
Comment (halseth) [inline filterdb/db.go]: style: add empty comment line above `NOTE`

File: filterdb/db.go [hunk 4]
```diff
@@ -107,6 +110,29 @@ func New(db walletdb.DB, params chaincfg.Params) (*FilterStore, error) {
...
+				return err
+			}
+			if _, err := filters.CreateBucket(regBucket); err != nil {
+				return err
+			}
+		default:
+			return fmt.Errorf("unknown filter type: %v", fType)
+		}
+
+		return nil
```
Comment (Crypt-iQ) [inline filterdb/db.go]: This statement can't be reached, any way to coalesce this with the above default case?

Comment (halseth) [reply filterdb/db.go]: Looks okay, as it will fall through after succesfully deleting the bucket for the filter type above

Comment (Crypt-iQ) [reply filterdb/db.go]: Ah you're right

**Pull request merged**